### PR TITLE
fix(core): ai replace selection

### PR DESCRIPTION
**Pull request by yoyoyohamapi**
### TL;DR

* Fix the issue of inaccurate content replacement in AI Replace Selection
* Optimize unit Tests utils

### What Changed
1. Fixed the issue of inaccurate content replacement in AI Replace Selection:
  - Convert the AI Answer into a Snapshot, then transform it into a sequence of Blocks ready for insertion.
   - Invoke the `replaceSelectedTextWithBlocks` command to replace the current selection with blocks (given the complexity of block combinations, this command uses [ts-pattern](https://github.com/gvergnaud/ts-pattern) implementation to ensure compile-time prevention of pattern handling omissions).
2. Optimized unit test assertions for commands, now allowing direct document content comparison using `toEqualDoc`.
```ts
const host = affine`
  <affine-page id="page">
    <affine-note id="note">
      <affine-paragraph id="paragraph-1">Hel<anchor />lo</affine-paragraph>
      <affine-paragraph id="paragraph-2">Wor<focus />ld</affine-paragraph>
    </affine-note>
  </affine-page>
`;

// ....

const expected = affine`
  <affine-page id="page">
    <affine-note id="note">
      <affine-paragraph id="paragraph-1">Hel111</affine-paragraph>
      <affine-code id="code"></affine-code>
      <affine-paragraph id="paragraph-2">222ld</affine-paragraph>
    </affine-note>
  </affine-page>
`;

expect(host.doc).toEqualDoc(expected.doc);
```
3. Added support for text cursors in unit test template syntax.

## Summary by CodeRabbit

> CLOSE BS-3278

- **New Features**
  - Introduced the ability to replace selected text in documents with blocks, supporting advanced merging and insertion scenarios for various block types.
- **Bug Fixes**
  - Improved handling of text selection and cursor placement in document templates used for testing.
- **Tests**
  - Added comprehensive tests for replacing selected text with blocks, including edge cases and complex selection scenarios.
  - Enhanced test utilities for document structure comparison and selection handling.
  - Updated end-to-end tests to verify correct replacement of selected text via AI-driven actions.
- **Chores**
  - Added and updated dependencies to support new features.

### CODE REVIEW

**Comment by yoyoyohamapi** [comment]
* **#11875** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/11875?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/11875?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `canary`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/toeverything?ref=merge-queue-instructions-in-stack-comment&prId=11875).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/frontend/core/src/blocksuite/ai/utils/editor-actions.ts [hunk 1]
```diff
@@ -109,19 +113,44 @@ export const replace = async (
   );
 
   if (textSelection) {
-    host.std.command.exec(deleteTextCommand, { textSelection });
+    const fragmentDoc = host.doc.workspace.createDoc('REPLACEMENT_TEMP_DOC');
```
Comment (yoyoyohamapi) [inline packages/frontend/core/src/blocksuite/ai/utils/editor-actions.ts]: For snapshot to blocks, I need to create a temp doc(then remove it), is there a better way to do it? @Saul-Mirone

Comment (Saul-Mirone) [reply packages/frontend/core/src/blocksuite/ai/utils/editor-actions.ts]: maybe create a temporary workspace to do that is better.

File: tests/affine-cloud-copilot/e2e/chat-with/text.spec.ts [hunk 2]
```diff
@@ -62,13 +62,16 @@ test.describe('AIChatWith/Text', () => {
     loggedInPage: page,
     utils,
   }) => {
-    const { translate } = await utils.editor.askAIWithText(page, 'Apple');
-    const { answer } = await translate('German');
-    await expect(answer).toHaveText(/Apfel/, { timeout: 10000 });
+    const { fixSpelling } = await utils.editor.askAIWithText(
+      page,
+      'I Loev Apple'
+    );
```
Comment (akumatus) [inline tests/affine-cloud-copilot/e2e/chat-with/text.spec.ts]: Select only `Loev` and then fix the spelling?

File: blocksuite/affine/shared/src/commands/model-crud/replace-selected-text-with-blocks.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+  const { startBlockModel, startOffset } = result;
+
+  const parent = doc.getParent(startBlockModel.id);
+  if (!parent) return next();
+
+  const pattern = getBlocksPattern(blocks);
+  const index = parent.children.findIndex(x => x.id === startBlockModel.id);
+
+  match(pattern)
+    .with({ first: 'paragraph', multiple: false }, () => {
```
Comment (akumatus) [inline blocksuite/affine/shared/src/commands/model-crud/replace-selected-text-with-blocks.ts]: Do we just need to check whether it is a paragraph? What about list and callout blocks?

File: tests/affine-cloud-copilot/e2e/chat-with/text.spec.ts (hunk 2 above)
Comment (yoyoyohamapi) [reply tests/affine-cloud-copilot/e2e/chat-with/text.spec.ts]: Let me c.... how to select specific text within a paragraph.

File: blocksuite/affine/shared/src/commands/model-crud/replace-selected-text-with-blocks.ts (hunk 3 above)
Comment (yoyoyohamapi) [reply blocksuite/affine/shared/src/commands/model-crud/replace-selected-text-with-blocks.ts]: Optimized:

* `canMergeBlocks`：Determines if blocks can be merged based on their types
* `isMergableBlock`: Check if a block is mergeable in general

> `canMergeBlocks` & `isMergableBlock` should be decoupled with flavor in the future

File: tests/affine-cloud-copilot/e2e/chat-with/text.spec.ts (hunk 2 above)
Comment (yoyoyohamapi) [reply tests/affine-cloud-copilot/e2e/chat-with/text.spec.ts:68]: Playwright currently can't handle this

File: packages/frontend/core/src/blocksuite/ai/utils/editor-actions.ts [hunk 4]
```diff
@@ -109,19 +114,52 @@ export const replace = async (
...
-      host
-    );
-    if (snapshot) {
-      await transformer.snapshotToSlice(
-        snapshot,
-        host.doc,
-        firstBlockParent.model.id,
-        firstIndex + 1
+    const collection = new WorkspaceImpl({
+      id: 'SLIDES_PREVIEW',
```
Comment (Saul-Mirone) [inline packages/frontend/core/src/blocksuite/ai/utils/editor-actions.ts]:
```suggestion
      id: 'AI_REPLACE',
```

File: blocksuite/affine/shared/src/__tests__/utils.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { Doc as YDoc } from 'yjs';
+
+export function createTestDoc() {
+  return new YDoc();
+}
```
Comment (akumatus) [inline blocksuite/affine/shared/src/__tests__/utils.ts]: There is no place to use createTestDoc?

File: packages/frontend/core/src/blocksuite/ai/utils/editor-actions.ts [hunk 6]
```diff
@@ -109,19 +114,52 @@ export const replace = async (
...
-    );
-    if (snapshot) {
-      await transformer.snapshotToSlice(
-        snapshot,
-        host.store,
-        firstBlockParent.model.id,
-        firstIndex + 1
+    const collection = new WorkspaceImpl({
+      id: 'AI_REPLACE',
+    });
```
Comment (akumatus) [inline packages/frontend/core/src/blocksuite/ai/utils/editor-actions.ts:121]: Why need to create a new workspace?

File: blocksuite/affine/shared/src/__tests__/helpers/affine-template.unit.spec.ts [hunk 7]
```diff
@@ -80,4 +81,79 @@ describe('helpers/affine-template', () => {
       `;
     }).toThrow();
   });
+
+  it('should handle text selection with anchor and focus', () => {
+    const host = affine`
```
Comment (Saul-Mirone) [inline blocksuite/affine/shared/src/__tests__/helpers/affine-template.unit.spec.ts:86]: Do you plan to replace it with tsx in this pr or next pr?

File: blocksuite/affine/shared/src/__tests__/helpers/affine-test-utils.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import type { BlockModel, Store } from '@blocksuite/store';
+import { expect } from 'vitest';
+
+declare module 'vitest' {
```
Comment (Saul-Mirone) [inline blocksuite/affine/shared/src/__tests__/helpers/affine-test-utils.ts:4]: Maybe have a `@blocksuite/test-kit` package is more helpful than put it in `@blocksuite/affine-shared`?

File: blocksuite/affine/shared/src/__tests__/helpers/affine-template.unit.spec.ts (hunk 7 above)
Comment (yoyoyohamapi) [reply blocksuite/affine/shared/src/__tests__/helpers/affine-template.unit.spec.ts:86]: Yeah, it's a future project(RFC --> Test Framework).

File: blocksuite/affine/shared/src/__tests__/utils.ts (hunk 5 above)
Comment (yoyoyohamapi) [reply blocksuite/affine/shared/src/__tests__/utils.ts]: removed

**Review by akumatus — APPROVED** [review]
